### ISSUE TITLE
RFC: Uniform Function Call Syntax (UFCS)

### ISSUE THREAD

**Issue by ajor** [bpftrace/bpftrace#3499]
Uniform Function Call Syntax (UFCS) is a language feature which allows any function to be called using the syntax for methods calls.

References:
- https://en.wikipedia.org/wiki/Uniform_Function_Call_Syntax
- https://tour.dlang.org/tour/en/gems/uniform-function-call-syntax-ufcs

It provides a simple way to implement "methods" (#3361), without turning BpfScript into an object oriented language. It maintains the pros mentioned in that issue (the main one for me being readability), while resolving most of the cons.

This syntax is compatible with both builtin functions and user defined functions, and can be used as we start to build out a standard library.

**User Defined Functions**
```
fn starts_with($str: string, $pattern: string): bool

starts_with(comm, "/usr/local/bin/");  // Traditional
comm.starts_with("/usr/local/bin/");   // UFCS
```

**Builtin Functions**
```
@map[1,2,3] = "hi";

delete($map: typeof(@map), $key: (int64, int64, int64)): void

delete(@map, (1, 2, 3));  // Traditional
@map.delete((1, 2, 3));   // UFCS
```

`@map.delete(1, 2, 3)` would be my ideal syntax, but we can work towards that from here.

**Chaining**
```
fn substr($str: string, $start_idx: uint32, $end_idx: uint32): string
fn contains($str: string, $pattern: string): bool

contains(substr($mystr, 10, 20), "python");  // Traditional
$mystr.substr(10, 20).contains("python")     // UFCS
```

**Comment by jordalgo** [comment]
Will functions in external c libraries also have this syntax available to them?

**Comment by ajor** [comment]
Yeah, any kind of function can be called either with the traditional syntax or UFCS.

**Comment by viktormalik** [comment]
I generally like this for the already stated purposes (readability, chaining, etc.) but we must realize that allowing this for all functions will allow things like:
```
"pid: %d\n".printf(pid);
4.bswap();
"vfs_read".kaddr();
4.sizeof();
(struct task_struct).offsetof(pid);   # this would probably not pass the parser
```
and, if we allow UFCS for parameters which are "enums":
```
monotonic.nsecs()
bpftrace.ustack()
```
and so on.

We may want to discuss if we allow UFCS for all types.

**Comment by danobi** [comment]
I suspect unrestricted UFCS is likely to enable just as much confusing syntax as it would enable "cleaner" syntax. Can we collect stats on how many cases where functions could be chained and compare with number of places where UFCS does not enhance readability (ie. function call with single arg)?

Restricting UFCS on the other hand starts to get complicated. I'm not sure it's simpler at all - it seems to me much more like bound/unbound methods.

**Comment by ajor** [comment]
> I generally like this for the already stated purposes (readability, chaining, etc.) but we must realize that allowing this for all functions will allow things like:
> 
> ```
> "pid: %d\n".printf(pid);
> 4.bswap();
> "vfs_read".kaddr();
> 4.sizeof();
> (struct task_struct).offsetof(pid);   # this would probably not pass the parser
> ```

I'd say that anything that can be parsed as `expr.ident(vargs)` should be valid. i.e. all of these except the `offsetof`. Some of them do look odd, but if users want to write their scripts that way I don't see a reason to stop them.

> and, if we allow UFCS for parameters which are "enums":
> 
> ```
> monotonic.nsecs()
> bpftrace.ustack()
> ```

This is a good point - trickier situation here because these are magical idents that only have meaning within their builtin function calls. Even so, if these get translated internally into `nsecs(monotonic)`, is there any harm in allowing it?

> I suspect unrestricted UFCS is likely to enable just as much confusing syntax as it would enable "cleaner" syntax.

Agreed that misuse of UFCS could look pretty ugly. The standard library and tools wouldn't do anything like that, so nasty examples would be limited to people's custom scripts, which only they are likely to look at. Is there much harm in that? We're giving users a tool, it's up to them how they want to use it.

> Can we collect stats on how many cases where functions could be chained and compare with number of places where UFCS does not enhance readability (ie. function call with single arg)?

There's an infinite number of possibilities in how user defined functions could be chained so I don't think it's possible to collect any kind of stats. Also chaining isn't the only benefit of UFCS - or even primary benefit in my mind. I prefer reading something like `if ("mylongstr".contains("long"))` over `if (contains("mylongstr", "long"))` as it feels more like reading a sentence and is clearer about the order of arguments.

**Comment by batu3H** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.
